### PR TITLE
Universal Sierra compiler

### PR DESCRIPTION
**Pull request by marioiordanov**
## Usage related changes

Introduce universal Sierra compiler.

## Development related changes

All cairo_lang_starknet_classes::casm_contract_class::CasmContractClass::from_contract_class are replaced with usc::compile_contract.
Removed compute_casm_class_hash function and use starknet-rs implementation of CompiledClass

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] No unused dependencies - `./scripts/check_unused_deps.sh`
- [x] Performed code self-review
- [ ] Rebased to the last commit of the target branch (or merged it into my branch)
- [ ] Documented the changes
- [ ] Linked the issues which this PR resolves
- [ ] Checked the TODO section in README.md if this PR resolves it
- [ ] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet-devnet-core/src/state/mod.rs [hunk 1]
```diff
@@ -284,11 +284,19 @@ impl CustomState for StarknetState {
...
+                    .map_err(|err| Error::SerializationError { origin: err.to_string() })?,
+            )
+            .map_err(|_| Error::SierraCompilationError)?;
+
+            let casm_hash = Felt::from(
+                serde_json::from_value::<CompiledClass>(casm_json)
+                    .map_err(|err| Error::DeserializationError { origin: err.to_string() })?
+                    .class_hash()
+                    .map_err(|err| Error::UnexpectedInternalError { msg: err.to_string() })?,
+            );
```
Comment (FabijanC) [inline crates/starknet-devnet-core/src/state/mod.rs:298]: Perhaps extract to a `casm_json_to_hash` function.

File: crates/starknet-devnet-core/src/utils.rs [hunk 2]
```diff
@@ -112,15 +108,14 @@ pub(crate) mod test_utils {
...
-        )
-        .unwrap();
+        let casm_contract_class_json =
+            usc::compile_contract(serde_json::to_value(contract_class.clone()).unwrap()).unwrap();
+
+        let compiled_class_hash = serde_json::from_value::<CompiledClass>(casm_contract_class_json)
+            .unwrap()
+            .class_hash()
+            .unwrap()
+            .into();
```
Comment (FabijanC) [inline crates/starknet-devnet-core/src/utils.rs:118]: Perhaps the suggested `casm_json_to_hash` could be used in tests as well, like here and in other similar places.

File: crates/starknet-devnet-core/src/starknet/state_update.rs [hunk 3]
```diff
@@ -47,13 +46,15 @@ mod tests {
         let sierra_class_hash =
             ContractClass::Cairo1(contract_class.clone()).generate_hash().unwrap();
 
-        let casm_contract_class = CasmContractClass::from_contract_class(
-            contract_class.clone(),
-            true,
-            MAX_BYTECODE_SIZE_LIMIT,
```
Comment (FabijanC) [inline crates/starknet-devnet-core/src/starknet/state_update.rs:53]: Should the constant `MAX_BYTECODE_SIZE_LIMIT` be removed now that it's unused?

File: Cargo.toml [hunk 4]
```diff
@@ -61,33 +61,34 @@ rand = "0.8.5"
 rand_chacha = "0.3.1"
 rand_mt = "4.2.2"
 regex_generate = "0.2.3"
+usc = { git = "https://github.com/software-mansion/universal-sierra-compiler", rev = "581402f", package = "universal-sierra-compiler" }
```
Comment (FabijanC) [inline Cargo.toml:64]: It's been released as a crate: https://spaceshard.slack.com/archives/C03031Y0LKC/p1711116989370019

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: crates/starknet-devnet-core/src/state/mod.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet-devnet-core/src/state/mod.rs]: [b05f918](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/391/commits/b05f9187a30724d91fd9b78d16c59fc53a9ea601)

File: crates/starknet-devnet-core/src/starknet/state_update.rs (hunk 3 above)
Comment (marioiordanov) [reply crates/starknet-devnet-core/src/starknet/state_update.rs:53]: [90c2edf](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/391/commits/90c2edf13ffd8d2370cc41eaf1e4a61247504828)

File: Cargo.toml (hunk 4 above)
Comment (marioiordanov) [reply Cargo.toml]: [67fc2be](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/391/commits/67fc2be0d9acbabd62c26789f9f6a4fb0703f4bd)

File: .circleci/config.yml [hunk 5]
```diff
@@ -63,11 +63,11 @@ jobs:
             - target/release/deps
           key: << parameters.cargo_cache_key >>
       - run:
-          name: Test
-          # if jobs not limited - fails
-          # problem: https://app.circleci.com/pipelines/github/0xSpaceShard/starknet-devnet-rs/339/workflows/97e98c29-1563-4aa0-b716-4bfd023c563e/jobs/335/steps
-          # solution: https://stackoverflow.com/questions/71962406/how-to-use-less-memory-when-compiling-to-avoid-killing-the-build
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply .circleci/config.yml]: message returned

File: crates/starknet-devnet-core/src/utils.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet-devnet-core/src/utils.rs]: [b05f918](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/391/commits/b05f9187a30724d91fd9b78d16c59fc53a9ea601)

**Review by FabijanC — APPROVED** [review]

**Comment by FabijanC** [comment]
note: cannot be merged because it's not using a stable starknet-rs version so it can't be released on crates.io

**Review by FabijanC — COMMENTED** [review]
There are still some merge conflicts. Also, I updated starknet-rs in #396 (to the same versions as this PR), but whichever gets merged sooner is ok

**Comment by marioiordanov** [comment]
closing in favour of #403

**Pull request closed without merging**